pipeline/mutate/mutator_id_token.go

```diff
@@ -40,47 +42,61 @@ type MutatorIDTokenRegistry interface {
 	credentials.SignerRegistry
 }
 
-type CredentialsIDTokenConfig struct {
-	Audience []string `json:"aud"`
-}
-
 type MutatorIDToken struct {
 	c configuration.Provider
 	r MutatorIDTokenRegistry
+	t *template.Template
+}
+
+type CredentialsIDTokenConfig struct {
+	Claims jwt.MapClaims `json:"claims"`
 }
 
 func NewMutatorIDToken(c configuration.Provider, r MutatorIDTokenRegistry) *MutatorIDToken {
-	return &MutatorIDToken{r: r, c: c}
+	return &MutatorIDToken{r: r, c: c, t: newTemplate("id_token")}
 }
 
 func (a *MutatorIDToken) GetID() string {
 	return "id_token"
 }
 
-func (a *MutatorIDToken) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, _ pipeline.Rule) error {
+func (a *MutatorIDToken) WithCache(t *template.Template) {
+	a.t = t
+}
+
+func (a *MutatorIDToken) Mutate(r *http.Request, session *authn.AuthenticationSession, config json.RawMessage, rl pipeline.Rule) error {
 	if len(config) == 0 {
 		config = []byte("{}")
 	}
 
-	var cc CredentialsIDTokenConfig
-	d := json.NewDecoder(bytes.NewBuffer(config))
-	d.DisallowUnknownFields()
-	if err := d.Decode(&cc); err != nil {
-		return errors.WithStack(err)
-	}
+	var tmpl *template.Template
+	var err error
 
-	now := time.Now().UTC()
-	claims := jwt.MapClaims{}
-	if session.Extra != nil {
-		for k, v := range session.Extra {
-			claims[k] = v
+	templateID := rl.GetID()
+	tmpl = a.t.Lookup(templateID)
```

Comment:
It makes sense to define the variable where it's actually being used/defined - this prevents potential nil panics in future code changes.

```suggestion
	tmpl := a.t.Lookup(templateID)
```